extensions/amp-stream-gallery/1.0/stream-gallery.js

```diff
@@ -0,0 +1,217 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {BaseCarousel} from '../../amp-base-carousel/1.0/base-carousel';
+import {
+  useCallback,
+  useLayoutEffect,
+  useMemo,
+  useRef,
+  useState,
+} from '../../../src/preact';
+import {useStyles} from './stream-gallery.jss';
+
+const DEFAULT_MEASUREMENT = {
+  visibleCount: 1,
+  maxContainerWidth: Number.MAX_VALUE,
+};
+
+/**
+ * @param {!StreamGalleryDef.Props} props
+ * @return {PreactDef.Renderable}
+ */
+export function StreamGallery({
+  arrowPrev: customArrowPrev,
+  arrowNext: customArrowNext,
+  children,
+  extraSpace,
+  insetArrowVisibility,
+  loop,
+  maxItemWidth = Number.MAX_VALUE,
+  minItemWidth = 1,
+  maxVisibleCount = Number.MAX_VALUE,
+  minVisibleCount = 1,
+  outsetArrows,
+  peek = 0,
+  snap,
+  style,
+  ...rest
+}) {
+  const galleryRef = useRef(null);
+  const [measurements, setMeasurements] = useState(DEFAULT_MEASUREMENT);
+  const arrowPrev = useMemo(
+    () =>
+      customArrowPrev || <DefaultArrow by={-1} outsetArrows={outsetArrows} />,
+    [customArrowPrev, outsetArrows]
+  );
+  const arrowNext = useMemo(
+    () =>
+      customArrowNext || <DefaultArrow by={1} outsetArrows={outsetArrows} />,
+    [customArrowNext, outsetArrows]
+  );
+
+  const measure = useCallback(
+    (containerWidth) =>
+      getVisibleCount(
+        maxItemWidth,
+        minItemWidth,
+        maxVisibleCount,
+        minVisibleCount,
+        children.length,
+        peek,
+        containerWidth
+      ),
+    [
+      maxItemWidth,
+      minItemWidth,
+      maxVisibleCount,
+      minVisibleCount,
+      children.length,
+      peek,
+    ]
+  );
+
+  // Adjust visible slide count when container size or parameters change.
+  useLayoutEffect(() => {
```

Comment:
I believe there is a subtle hidden bug in this layoutEffect hook.
Namely that changes to the `galleryRef.current` pointer would not retrigger the effect.
If it is working, it is because we are getting lucky with the synchronization of the ref setting and when the render is happening.

The solution is to use a [callback ref](https://reactjs.org/docs/hooks-faq.html#how-can-i-measure-a-dom-node).

In this case, I think that means:
```js
const setupMeasureRef = useCallback(() => { ...  }, [measure]);
return <div ref={setupMeasureRef}> ... </div>;
```

Reply:
I just removed the outer div with the ref in favor of using one provided by the `BaseCarousel`, does this comment still apply?

Reply:
The issue that `useLayoutEffect` will not be notified about a change to `ref.current.node` still exists. The way the code is currently written, I don't think it would be a problem since the nodes on which the refs exist are static (BaseCarousel / ScrollerRef).

If in the future the ref moves to a component or node that is in any way rendered conditionally, then it becomes a bug

Reply:
The DOM ref changes is a big headache in React. They are not directly supported by any API. I think the overall idea is, as much as possible, to assume DOM refs do not change unless it's an explicitly supported case.